src/Analyzers/CSharp/CodeFixes/UsePrimaryConstructor/CSharpUsePrimaryConstructorCodeFixProvider.cs

```diff
@@ -166,10 +173,46 @@ private static async Task UsePrimaryConstructorAsync(
                         .WithoutLeadingTrivia()
                         .WithTrailingTrivia(triviaAfterName)
                         .WithAdditionalAnnotations(Formatter.Annotation));
+
+                return WithCommentMoved(finalTypeDeclaration);
             });
 
         return;
 
+        TypeDeclarationSyntax WithCommentMoved(TypeDeclarationSyntax finalTypeDeclaration)
+        {
+            var firstMember = typeDeclaration.Members.First();
+            if (firstMember == constructorDeclaration || removedMembers.Any(kvp => kvp.Value.memberNode == firstMember))
+            {
+                // We're removing the first member in the type.  If this member had comments above it (with a blank line
+                // between it and the member) then keep those comments around.
+                var triviaToMove = GetLeadingCommentTrivia(firstMember);
+                if (triviaToMove.Length > 0)
+                {
+                    var nextToken = finalTypeDeclaration.OpenBraceToken.GetNextToken();
+                    return finalTypeDeclaration.ReplaceToken(
+                        nextToken,
+                        nextToken.WithPrependedLeadingTrivia(triviaToMove));
+                }
+            }
+
+            return finalTypeDeclaration;
+        }
+
+        ImmutableArray<SyntaxTrivia> GetLeadingCommentTrivia(MemberDeclarationSyntax firstMember)
+        {
+            var leadingTrivia = firstMember.GetLeadingTrivia().ToImmutableArray();
+
+            for (var i = leadingTrivia.Length - 1; i >= 0; i--)
```

Comment:
>for (var i = leadingTrivia.Length - 1; i >= 0; i--) [](http://example.com/codeflow?start=12&length=51)

It seems odd to me that we continually keep checking for matches for each leadingTrivia prefix, even if the first trivia doesn't match. The Matcher stuff looks pretty nifty, but I'm not sure it fits for me in this context.

Reply:
What would be your preference instead?

Reply:
I probably have a couple off by one issues here, but I had just thought expanding it out to something like the following wouldn't really complicate things too bad:

```
        ImmutableArray<SyntaxTrivia> GetLeadingCommentTrivia(MemberDeclarationSyntax firstMember)
        {
            var leadingTrivia = firstMember.GetLeadingTrivia().ToImmutableArray();

            if (leadingTrivia.Length < 3 ||
                !leadingTrivia[0].IsSingleOrMultiLineComment() ||
                !leadingTrivia[1].IsKind(SyntaxKind.EndOfLineTrivia))
            {
                return [];
            }

            int i;
            for (i = 2; i < leadingTrivia.Length - 1; i++)
            {
                if (leadingTrivia[i].Kind() != SyntaxKind.WhitespaceTrivia)
                    break;
            }

            if (!leadingTrivia[i].IsKind(SyntaxKind.EndOfLineTrivia))
                return [];

            return leadingTrivia[..(i + 1)];
        }
```

Reply:
> even if the first trivia doesn't match.

Not sure what is meant by this. We're walking backwards to find the *last* preceding comment that is followed by a blank line.

Reply:
I was concerned about the case where there is a large number of leading trivia. We would reverse iterate through each index and do a TryMatch. Each TryMatch call essentially needs to check all matches until it finds a mismatch, if I'm understanding properly. This feels a bit more effort than is necessary, as it feels like we can just forward iterate through and check the matches once.

Reply:
Not quite. 

The matching function is doing what your loops are doing, just through a declarative API. They're not looking for all matches, they're just checking if the index they are at are a match, no more.

Reply:
> I probably have a couple off by one issues here, but I had just thought expanding it out to something like the following wouldn't really complicate things too bad:
> 
> ```
>         ImmutableArray<SyntaxTrivia> GetLeadingCommentTrivia(MemberDeclarationSyntax firstMember)
>         {
>             var leadingTrivia = firstMember.GetLeadingTrivia().ToImmutableArray();
> 
>             if (leadingTrivia.Length < 3 ||
>                 !leadingTrivia[0].IsSingleOrMultiLineComment() ||
>                 !leadingTrivia[1].IsKind(SyntaxKind.EndOfLineTrivia))
>             {
>                 return [];
>             }
> 
>             int i;
>             for (i = 2; i < leadingTrivia.Length - 1; i++)
>             {
>                 if (leadingTrivia[i].Kind() != SyntaxKind.WhitespaceTrivia)
>                     break;
>             }
> 
>             if (!leadingTrivia[i].IsKind(SyntaxKind.EndOfLineTrivia))
>                 return [];
> 
>             return leadingTrivia[..(i + 1)];
>         }
> ```

This is what the matcher is doing. Just without the need for keeping track of all the indices, and without needing all those loops. 

Think of it as creating a regex. Just over SyntaxTrivia, not chars. And that you give the starting point to try to match once at, and it returns if it succeeded and what the end point is.

Reply:
But if it's not a match, it will decrement, and redo the check again against a slightly smaller group of trivia, no?

Reply:
No. It's not testing against the entire group. It's checking if it's a match at that particular index. Just like your loop is doing. You're walking from one end to other seeing if the location you are at is a match. That's what my loop does too. With the matcher being told the index to do the check at. 

The matcher doesn't walk the list itself looking for the first match. It checks if it is *at* a match and returns that info. The caller does the walking.

Reply:
Chatted offline, and Cyrus helped fix where my mental model was wrong. The loop backwards basically ensures that the WS* check will only be done after the passed in index is verified to be a comment, so my concern about it happening gobs of times is wrongly formed. Also, my example code was wrong because it wasn't looping backwards to find the last entry.